gcc/config/riscv/rvp.md

```diff
@@ -0,0 +1,7058 @@
+		  (sign_extend:SI (vec_select:HI (match_dup 3) (parallel [(const_int 3)]))))
+		(mult:SI
+		  (sign_extend:SI (vec_select:HI (match_dup 2) (parallel [(const_int 3)])))
+		  (sign_extend:SI (vec_select:HI (match_dup 3) (parallel [(const_int 2)]))))))))]
+  "TARGET_ZPN && TARGET_64BIT"
+  "smslxda\t%0, %2, %3"
+  [(set_attr "type" "dsp64")])
+
+;; SMSR64, UMSR64
+(define_insn "<su>msr64"
+  [(set (match_operand:DI 0 "register_operand"       "=r")
+	(minus:DI
+	  (match_operand:DI 1 "register_operand"     " 0")
+	  (mult:DI
+	    (any_extend:DI
+	      (match_operand:SI 2 "register_operand" " r"))
+	    (any_extend:DI
+	      (match_operand:SI 3 "register_operand" " r")))))]
+  "TARGET_ZPSF && !TARGET_64BIT"
+  "<su>msr64\t%0, %2, %3"
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+(define_insn "v<su>msr64"
+  [(set (match_operand:DI 0 "register_operand"             "=r")
+	(minus:DI
+	  (minus:DI
+	  (match_operand:DI 1 "register_operand"    " 0")
+	    (mult:DI
+	      (any_extend:DI
+		(vec_select:SI
+		  (match_operand:V2SI 2 "register_operand" " r")
+		  (parallel [(const_int 0)])))
+	      (any_extend:DI
+		(vec_select:SI
+		  (match_operand:V2SI 3 "register_operand" " r")
+		  (parallel [(const_int 0)])))))
+	    (mult:DI
+	      (any_extend:DI
+		(vec_select:SI (match_dup 2) (parallel [(const_int 1)])))
+	      (any_extend:DI
+		(vec_select:SI (match_dup 3) (parallel [(const_int 1)]))))))]
+  "TARGET_ZPSF && TARGET_64BIT"
+  "<su>msr64\t%0, %2, %3"
+  [(set_attr "type" "dsp64")
+   (set_attr "mode" "DI")])
+
+;; SMUL 8|16, SMULX 8|16, UMUL 8|16, UMULX 8|16
+;; SMUL8
+(define_insn "smul8"
+  [(set (match_operand:DI 0 "register_operand"             "=r")
+	(unspec:DI [(match_operand:SI 1 "register_operand" " r")
+		    (match_operand:SI 2 "register_operand" " r")]
+		    UNSPEC_SMUL8))]
+  "TARGET_ZPSF"
+  "smul8\t%0, %1, %2"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "V4HI")])
+
+;; SMULX8
+(define_insn "smulx8"
+  [(set (match_operand:DI 0 "register_operand"             "=r")
+	(unspec:DI [(match_operand:SI 1 "register_operand" " r")
+		    (match_operand:SI 2 "register_operand" " r")]
+		    UNSPEC_SMULX8))]
+  "TARGET_ZPSF"
+  "smulx8\t%0, %1, %2"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "V4HI")])
+
+;; UMUL8
+(define_insn "umul8"
+  [(set (match_operand:DI 0 "register_operand"             "=r")
+	(unspec:DI [(match_operand:SI 1 "register_operand" " r")
+		    (match_operand:SI 2 "register_operand" " r")]
+		    UNSPEC_UMUL8))]
+  "TARGET_ZPSF"
+  "umul8\t%0, %1, %2"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "V4HI")])
+
+;; UMULX8
+(define_insn "umulx8"
+  [(set (match_operand:DI 0 "register_operand"             "=r")
+	(unspec:DI [(match_operand:SI 1 "register_operand" " r")
+		    (match_operand:SI 2 "register_operand" " r")]
+		    UNSPEC_UMULX8))]
+  "TARGET_ZPSF"
+  "umulx8\t%0, %1, %2"
+  [(set_attr "type" "simd")
+   (set_attr "mode" "V4HI")])
+
+;; S|U MUL16
```

Comment:
```c
return 0xFF797D61F26D4629 != __rv__smul16(0x01234567, 0x89ABCDEF);
```
Seems like this puts 31..0 to 63..32 and some extension to 63..32.
This returned 0xc0800000ff797d61, does not seem as a correct answer.